chromium/rules.js

```diff
@@ -403,45 +403,62 @@ RuleSets.prototype = {
    */
   potentiallyApplicableRulesets: function(host) {
     // Have we cached this result? If so, return it!
-    var cached_item = this.ruleCache.get(host);
-    if (cached_item !== undefined) {
-      log(DBUG, "Ruleset cache hit for " + host + " items:" + cached_item.length);
+    if (this.ruleCache.has(host)) {
+      let cached_item = this.ruleCache.get(host);
+      let len = cached_item ? cached_item.size : 0;
+      log(DBUG, "Ruleset cache hit for " + host + " items:" + len);
       return cached_item;
-    }
-    log(DBUG, "Ruleset cache miss for " + host);
-
-    var tmp;
-    var results = [];
-    if (this.targets.has(host)) {
-      // Copy the host targets so we don't modify them.
-      results = results.concat(this.targets.get(host));
+    } else {
+      log(DBUG, "Ruleset cache miss for " + host);
     }
 
     // Ensure host is well-formed (RFC 1035)
-    if (host.indexOf("..") != -1 || host.length > 255) {
+    if (host.length > 255 || host.indexOf('..') != -1) {
       log(WARN,"Malformed host passed to potentiallyApplicableRulesets: " + host);
       return null;
     }
 
+    // Let's search!!
+    let resultSet = new Set();
+
+    // Look for exact matches
+    if (this.targets.has(host)) {
+      let results = this.targets.get(host);
+      if (results) {
+        for (let result of results) {
+          resultSet.add(result);
+        }
+      }
+    }
+
     // Replace each portion of the domain with a * in turn
-    var segmented = host.split(".");
-    for (let i=0; i < segmented.length; i++) {
+    let segmented = host.split(".");
+    for (let i = 0; i < segmented.length; i++) {
       let tmp = segmented[i];
       segmented[i] = "*";
-      results = results.concat(this.targets.get(segmented.join(".")));
+
+      let results = this.targets.get(segmented.join('.'));
+      if (results) {
+        for (let result of results) {
+          resultSet.add(result);
+        }
+      }
+
       segmented[i] = tmp;
     }
+
     // now eat away from the left, with *, so that for x.y.z.google.com we
     // check *.z.google.com and *.google.com (we did *.y.z.google.com above)
-    for (var i = 2; i <= segmented.length - 2; ++i) {
-      var t = "*." + segmented.slice(i,segmented.length).join(".");
-      results = results.concat(this.targets.get(t));
+    for (let i = 2; i <= segmented.length - 2; ++i) {
+      let domain  = "*." + segmented.slice(i, segmented.length).join(".");
```

Comment:
This is unrelated to this PR, but just wanted to note for posterity. I think an algorithm for finding matching rulesets would be much simpler than this if they rulesets were structured as a prefix tree, where the domains are split into labels and reversed (assuming we can remove the RHS wildcards).